Refactor DissonanceFishNetPlayer class

YazanYahya98 opened the pull request:
Fixed a PlayerID null exception which led to proximity chat connection issues.

YazanYahya98 committed:
Refactor DissonanceFishNetPlayer class

Fixed a PlayerID null exception which led to proximity chat connection issues.

martindevans commented on Assets/Dissonance/Integrations/FishNet/DissonanceFishNetPlayer.cs:
```diff
@@ -8,48 +11,59 @@
-        public string PlayerId => _syncedPlayerName.Value;
+
+        public string _cachedPlayerId;
+        public string PlayerId => _cachedPlayerId ??= _syncedPlayerName.Value;
```
I don't think this is safe. It's not common, but it is possible for the player name to change, which would mean this cache was incorrect.

Is accessing `_syncedPlayerName.Value` so expensive that caching the result is worthwhile?

YazanYahya98 replied:
The _cachedPlayerId is used when untracking the player state via ManageTrackingStateCoroutine(false). Because DissonanceComms.StopTracking() relies on the current PlayerId property, we must temporarily use the cached value to ensure we are untracking the previous ID. This prevents the system from accidentally untracking the new ID (derived from _syncedPlayerName.Value) before it is fully applied.

martindevans replied:
aha I see, thanks for clarifying that 👍

martindevans commented on Assets/Dissonance/Integrations/FishNet/DissonanceFishNetPlayer.cs:
```diff
@@ -72,41 +87,93 @@ public override void OnOwnershipClient(NetworkConnection prevOwner)
-            if(!IsOwner) SetPlayerName(updatedName);
+            if (string.IsNullOrEmpty(updatedName) || updatedName.Equals(PlayerId, StringComparison.Ordinal))
+            {
+                Debug.LogWarning("[DissonanceFishNetPlayer] Received empty player name update!");
```
This warning looks like it could be misleading - if an update is received that's not null, but is the same as the current ID it will log:

> [DissonanceFishNetPlayer] Received empty player name update!

Which is not true!

Probably best to split this into two separate checks with different log messages.

YazanYahya98 replied:
You are right, I intended to delete the warning message (It was just for debugging). We can remove it and keep the return statement without any warnings.

martindevans replied:
I think keeping a warning of some kind makes sense, setting a null or empty name is quite weird.

YazanYahya98 replied:
Warning messages have been updated with a new commit

martindevans reviewed (commented):
Thanks for submitting this PR! I've attached a couple of comments with minor changes that should help in some unusual/edge cases.

YazanYahya98 committed:
Updated DissonanceFishNetPlayer script

Updated warning messages when verifying PlayerId at ManageTrackingStateCoroutine
